File: website/functions/_lib/session.ts
/**
 * 账户中心网页会话：AES-GCM 加密 cookie（stateless），web-design.md §6.2。
 *
 * cookie：httpOnly + Secure + SameSite=Lax，值为加密 payload（含 sub/资料/exp）。
 * 网页会话**不持久化 refresh_token**（账户中心无离线刷新需求，§6.1）。
 * 用加密 cookie 而非新 KV 命名空间：契约 §9.2 仅定义 HANDOFF/SESSIONS 两个命名空间。
 */
import type { Env } from "./env";
import { base64urlDecode, base64urlEncode, toBytes } from "./pkce";

export const SESSION_COOKIE = "lumen_session";
const DEFAULT_MAX_AGE = 60 * 60 * 8; // 8 小时

export interface WebSession {
  sub: string;
  display_name: string;
  avatar_url: string;
  exp: number; // epoch 秒（服务端权威过期时间）
}

/** 从 SESSION_ENC_KEY（base64）导入 AES-GCM 密钥（要求 32 字节）。 */
async function importKey(env: Env): Promise<CryptoKey> {
  const rawKey = decodeKey(env.SESSION_ENC_KEY);
  if (rawKey.length !== 32) {
    throw new Error("SESSION_ENC_KEY must decode to 32 bytes (AES-256-GCM)");
  }
  return crypto.subtle.importKey("raw", rawKey, { name: "AES-GCM" }, false, [
    "encrypt",
    "decrypt",
  ]);
}

/** 支持 base64 / base64url 的密钥解码。 */
function decodeKey(value: string): Uint8Array<ArrayBuffer> {
  const normalized = value.replace(/-/g, "+").replace(/_/g, "/");
  const pad = normalized.length % 4 === 0 ? "" : "=".repeat(4 - (normalized.length % 4));
  const binary = atob(normalized + pad);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

/** 加密会话为 cookie 值（iv.ciphertext，均 base64url）。 */
export async function sealSession(env: Env, session: WebSession): Promise<string> {
  const key = await importKey(env);
  const iv = toBytes(crypto.getRandomValues(new Uint8Array(12)));
  const plaintext = toBytes(new TextEncoder().encode(JSON.stringify(session)));
  const ciphertext = await crypto.subtle.encrypt({ name: "AES-GCM", iv }, key, plaintext);
  return `${base64urlEncode(iv)}.${base64urlEncode(new Uint8Array(ciphertext))}`;
}

/** 解密 cookie 值为会话；非法/篡改/过期返回 null。 */
export async function openSession(env: Env, cookieValue: string): Promise<WebSession | null> {
  const dot = cookieValue.indexOf(".");
  if (dot <= 0) return null;
  const ivPart = cookieValue.slice(0, dot);
  const ctPart = cookieValue.slice(dot + 1);
  let iv: Uint8Array<ArrayBuffer>;
  let ciphertext: Uint8Array<ArrayBuffer>;
  try {
    iv = base64urlDecode(ivPart);
    ciphertext = base64urlDecode(ctPart);
  } catch {
    return null;
  }
  if (iv.length !== 12) return null;
  try {
    const key = await importKey(env);
    const plaintext = await crypto.subtle.decrypt({ name: "AES-GCM", iv }, key, ciphertext);
    const session = JSON.parse(new TextDecoder().decode(plaintext)) as WebSession;
    if (typeof session.exp !== "number" || session.exp <= nowSeconds()) return null;
    return session;
  } catch {
    return null; // GCM 认证失败（篡改）或 JSON 损坏
  }
}

/** 读取请求 cookie 中的会话值。 */
export function readSessionCookie(request: Request): string | null {
  const header = request.headers.get("cookie");
  if (!header) return null;
  for (const part of header.split(";")) {
    const [name, ...rest] = part.trim().split("=");
    if (name === SESSION_COOKIE) return rest.join("=");
  }
  return null;
}

/** 构造 Set-Cookie（httpOnly + Secure + SameSite=Lax + Path=/）。 */
export function buildSessionCookie(value: string, maxAgeSeconds = DEFAULT_MAX_AGE): string {
  return [
    `${SESSION_COOKIE}=${value}`,
    "Path=/",
    "HttpOnly",
    "Secure",
    "SameSite=Lax",
    `Max-Age=${maxAgeSeconds}`,
  ].join("; ");
}

/** 构造清除会话的 Set-Cookie（Max-Age=0）。 */
export function clearSessionCookie(): string {
  return [
    `${SESSION_COOKIE}=`,
    "Path=/",
    "HttpOnly",
    "Secure",
    "SameSite=Lax",
    "Max-Age=0",
  ].join("; ");
}

export function defaultSessionExp(maxAgeSeconds = DEFAULT_MAX_AGE): number {
  return nowSeconds() + maxAgeSeconds;
}

// --- 账户中心 OIDC 登录流程上下文（/auth/login → /auth/callback）---
// 用短 TTL 加密 cookie 暂存 verifier/state，避免为网页登录再开 KV 命名空间。

export const AUTH_FLOW_COOKIE = "lumen_auth_flow";
const AUTH_FLOW_MAX_AGE = 600; // 10 分钟

export interface AuthFlowContext {
  verifier: string; // OIDC PKCE verifier
  state: string; // OIDC state（回调校验）
  exp: number;
}

export async function sealAuthFlow(env: Env, ctx: AuthFlowContext): Promise<string> {
  return sealSession(env, ctx as unknown as WebSession);
}

export async function openAuthFlow(
  env: Env,
  cookieValue: string,
): Promise<AuthFlowContext | null> {
  const opened = (await openSession(env, cookieValue)) as unknown as AuthFlowContext | null;
  if (!opened || typeof opened.verifier !== "string" || typeof opened.state !== "string") {
    return null;
  }
  return opened;
}

export function readAuthFlowCookie(request: Request): string | null {
  const header = request.headers.get("cookie");
  if (!header) return null;
  for (const part of header.split(";")) {
    const [name, ...rest] = part.trim().split("=");
    if (name === AUTH_FLOW_COOKIE) return rest.join("=");
  }
  return null;
}

export function buildAuthFlowCookie(value: string): string {
  return [
    `${AUTH_FLOW_COOKIE}=${value}`,
    "Path=/",
    "HttpOnly",
    "Secure",
    "SameSite=Lax",
    `Max-Age=${AUTH_FLOW_MAX_AGE}`,
  ].join("; ");
}

export function clearAuthFlowCookie(): string {
  return [`${AUTH_FLOW_COOKIE}=`, "Path=/", "HttpOnly", "Secure", "SameSite=Lax", "Max-Age=0"].join(
    "; ",
  );
}

export function defaultAuthFlowExp(): number {
  return nowSeconds() + AUTH_FLOW_MAX_AGE;
}

function nowSeconds(): number {
  return Math.floor(Date.now() / 1000);
}

File: server/internal/broker/session.go
package broker

import (
	"encoding/json"
	"net/http"
	"time"

	"lumen/internal/secure"
)

// Account-center cookies (decision 1). Both are host-only (NO Domain attribute),
// HttpOnly + Secure + SameSite=Lax + Path=/, sealed with AES-256-GCM under
// LUMEN_SESSION_ENC_KEY.
const (
	// sessionCookie holds the account-center web session (sub + profile).
	sessionCookie = "lumen_session"
	// authFlowCookie holds the short-lived OIDC login flow context (verifier +
	// state) for the web login (/auth/login → /auth/callback).
	authFlowCookie = "lumen_auth_flow"

	// sessionMaxAge is the web session lifetime (8h, mirrors session.ts).
	sessionMaxAge = 60 * 60 * 8
	// authFlowMaxAge is the login-flow context lifetime (10m, mirrors session.ts).
	authFlowMaxAge = 600
)

// webSession is the sealed account-center session payload (mirrors session.ts
// WebSession). exp is the server-authoritative expiry in epoch seconds.
type webSession struct {
	Sub         string `json:"sub"`
	DisplayName string `json:"display_name"`
	AvatarURL   string `json:"avatar_url"`
	Exp         int64  `json:"exp"`
}

// authFlowContext is the sealed OIDC login-flow context (mirrors session.ts
// AuthFlowContext). It carries the PKCE verifier and state across the redirect
// to the IdP and back.
type authFlowContext struct {
	Verifier string `json:"verifier"`
	State    string `json:"state"`
	Exp      int64  `json:"exp"`
}

// nowSeconds returns the current epoch second (server-authoritative time).
func nowSeconds() int64 { return time.Now().Unix() }

// --- account-center session (lumen_session) ---

// sealSession seals a webSession into a cookie value (mirrors session.ts
// sealSession).
func sealSession(sealer *secure.Sealer, s webSession) (string, error) {
	raw, err := json.Marshal(s)
	if err != nil {
		return "", err
	}
	return sealer.Seal(raw)
}

// openSession opens and validates a sealed session cookie value, returning nil
// on any malformed / tampered / expired value (mirrors session.ts openSession
// returning null).
func openSession(sealer *secure.Sealer, value string) *webSession {
	plain, err := sealer.Open(value)
	if err != nil {
		return nil
	}
	var s webSession
	if err := json.Unmarshal(plain, &s); err != nil {
		return nil
	}
	if s.Exp <= nowSeconds() {
		return nil
	}
	return &s
}

// readSessionCookie returns the raw sealed value of the session cookie, or ""
// (mirrors session.ts readSessionCookie).
func readSessionCookie(r *http.Request) string {
	return readCookie(r, sessionCookie)
}

// buildSessionCookie constructs a host-only Set-Cookie for the session
// (HttpOnly + Secure + SameSite=Lax + Path=/, NO Domain), mirroring session.ts
// buildSessionCookie.
func buildSessionCookie(value string) *http.Cookie {
	return newHostCookie(sessionCookie, value, sessionMaxAge)
}

// clearSessionCookie constructs a host-only Set-Cookie that clears the session
// (Max-Age=0), mirroring session.ts clearSessionCookie.
func clearSessionCookie() *http.Cookie {
	return newHostCookie(sessionCookie, "", 0)
}

// defaultSessionExp returns the authoritative session expiry (now + 8h).
func defaultSessionExp() int64 { return nowSeconds() + sessionMaxAge }

// --- OIDC login flow context (lumen_auth_flow) ---

// sealAuthFlow seals an authFlowContext into a cookie value (mirrors session.ts
// sealAuthFlow, which reuses sealSession).
func sealAuthFlow(sealer *secure.Sealer, ctx authFlowContext) (string, error) {
	raw, err := json.Marshal(ctx)
	if err != nil {
		return "", err
	}
	return sealer.Seal(raw)
}

// openAuthFlow opens a sealed auth-flow cookie value, returning nil on any
// malformed / tampered / expired value or when verifier/state are absent
// (mirrors session.ts openAuthFlow). The expiry guard reuses the same exp field
// as the session payload.
func openAuthFlow(sealer *secure.Sealer, value string) *authFlowContext {
	plain, err := sealer.Open(value)
	if err != nil {
		return nil
	}
	var ctx authFlowContext
	if err := json.Unmarshal(plain, &ctx); err != nil {
		return nil
	}
	if ctx.Verifier == "" || ctx.State == "" {
		return nil
	}
	if ctx.Exp <= nowSeconds() {
		return nil
	}
	return &ctx
}

// readAuthFlowCookie returns the raw sealed value of the auth-flow cookie, or ""
// (mirrors session.ts readAuthFlowCookie).
func readAuthFlowCookie(r *http.Request) string {
	return readCookie(r, authFlowCookie)
}

// buildAuthFlowCookie constructs a host-only Set-Cookie for the login flow
// (mirrors session.ts buildAuthFlowCookie).
func buildAuthFlowCookie(value string) *http.Cookie {
	return newHostCookie(authFlowCookie, value, authFlowMaxAge)
}

// clearAuthFlowCookie constructs a host-only Set-Cookie that clears the login
// flow (mirrors session.ts clearAuthFlowCookie).
func clearAuthFlowCookie() *http.Cookie {
	return newHostCookie(authFlowCookie, "", 0)
}

// defaultAuthFlowExp returns the login-flow context expiry (now + 10m).
func defaultAuthFlowExp() int64 { return nowSeconds() + authFlowMaxAge }

// --- shared cookie helpers ---

// newHostCookie builds a host-only cookie (decision 1): HttpOnly + Secure +
// SameSite=Lax + Path=/ and, crucially, NO Domain attribute so it is never sent
// to sibling subdomains. A negative maxAge (via Max-Age=0) clears the cookie.
func newHostCookie(name, value string, maxAge int) *http.Cookie {
	c := &http.Cookie{
		Name:     name,
		Value:    value,
		Path:     "/",
		HttpOnly: true,
		Secure:   true,
		SameSite: http.SameSiteLaxMode,
		// Domain intentionally left empty: host-only cookie (never Domain=.x.com).
	}
	if maxAge > 0 {
		c.MaxAge = maxAge
	} else {
		// Max-Age=0 clears the cookie. http.Cookie encodes MaxAge<0 as
		// "Max-Age=0", which is the exact clear semantics we want.
		c.MaxAge = -1
	}
	return c
}

// readCookie returns the value of the named cookie, or "" when absent.
func readCookie(r *http.Request, name string) string {
	c, err := r.Cookie(name)
	if err != nil {
		return ""
	}
	return c.Value
}
